File: java/solutions/src/main/java/io/github/ajoz/workshop/fp/java/part2/solutions/exercise_2/Exercise2.java
package io.github.ajoz.workshop.fp.java.part2.solutions.exercise_2;

import io.github.ajoz.workshop.fp.java.tools.Function2;

import java.util.List;

class Exercise2 {
    // Part 1:
    static Integer foo(final List<Integer> list,
                       final Integer initial,
                       final Function2<Integer, Integer, Integer> operator) {
        Integer accumulator = initial;
        for (final Integer element : list) {
            accumulator = operator.apply(accumulator, element);
        }

        return accumulator;
    }

    // Part 2:
    static <A, B> B bar(final List<A> list,
                        final B initial,
                        final Function2<B, A, B> operator) {
        B accumulator = initial;
        for (final A element : list) {
            accumulator = operator.apply(accumulator, element);
        }

        return accumulator;
    }

    // Part 3:
    static Integer sum(final List<Integer> list) {
        return bar(list, 0, (a, b) -> a + b);
    }

    static Integer product(final List<Integer> list) {
        return bar(list, 1, (a, b) -> a * b);
    }
}

File: kotlin/solutions/src/main/kotlin/io/github/ajoz/workshop/fp/part2/exercise2/Exercise2.kt
@file:Suppress("PackageName")

package io.github.ajoz.workshop.fp.part2.exercise2

// Part 1:
fun foo(list: List<Int>,
        initial: Int,
        operator: (Int, Int) -> Int): Int {
    var accumulator = initial
    for (element in list) {
        accumulator = operator(accumulator, element)
    }

    return accumulator
}

// Part 2:
fun <A, B> bar(list: List<A>,
               initial: B,
               operator: (B, A) -> B): B {
    // look at foo implementation and this implementation
    // isn't type inference a great thing :-)
    var accumulator = initial
    for (element in list) {
        accumulator = operator(accumulator, element)
    }

    return accumulator
}

// Part 3:
fun sum(list: List<Int>): Int {
    return bar(list, 0) { b, a ->
        b + a
    }
}

fun product(list: List<Int>): Int {
    return bar(list, 1) { b, a ->
        b * a
    }
}
